### ISSUE TITLE
Regression charging via dell usb-c dock

### ISSUE THREAD

**Issue by geraldhobbes** [Dasharo/dasharo-issues#587]
### Device

NovaCustom NV41

### Dasharo version

v1.7.1

### Affected component(s) or functionality

The charging functionality

### Brief summary

dell dock resets when attempting to charge

### How reproducible

The dock malfunctions 10 out of 10 attempts, except when:

1. charging with the DC-in jack
2. the battery percentage is high and the CPU has low load

When `/sys/class/power_supply/BAT0/capacity` said 97%, it switched to the different behavior.

### How to reproduce

1. Get a dell WD19 k20a001 docking station
2. Plug the dock into the left USB-C port of the laptop
3. Connect an external keyboard to the dock
5. Charge the laptop to 100% via the DC-in jack on the right, and keep it plugged in
4. Boot into DTS
6. In the linux shell, generate some cpu load with `sha256sum < /dev/urandom`
7. Observe how the external keyboard functions normally
8. Disconnect the DC-in jack

### Expected behavior

The external devices connected through the dock should keep functioning. The laptop should  draw power from the dock. This worked well in the v1.6.0 firmware.

### Actual behavior

The dock malfunctions:

   * the external keyboard stops working
   * the LED on the end of the dock's usb cable starts blinking
   * the HDMI and ethernet ports on the dock also stop working

### Screenshots

This video shows how the external keyboard stops working when disconnecting the power cord, so that the laptop tries to switch and draw power via the usb-c dock instead. It only starts to malfunction with artificial cpu load, because the battery is almost full.

https://github.com/Dasharo/dasharo-issues/assets/150605738/5ec70bd3-d70b-4983-876a-8ab6ac8c7f8b

### Additional context

When booted into an operating system, it notifies you about every second that a device was attached or removed when the dock malfunctions, while nothing is physically attached.
In the UEFI setup with the dock attached, the input via the internal keyboard had more latency then without the dock attached.

The behavior sounds similar to https://github.com/Dasharo/dasharo-issues/issues/431

### Solutions you've tried

An Anker USB-C dongle supporting power delivery that I had lying around worked as normal, so USB power delivery is not entirely broken, only with specific docks.

Downgrading the firmware in DTS with these commands solved the issue:

```sh
wget -O main.rom https://3mdeb.com/open-source-firmware/Dasharo/novacustom_nv4x_adl/v1.6.0/novacustom_nv4x_adl_v1.6.0.rom
wget -O ec.rom https://3mdeb.com/open-source-firmware/Dasharo/novacustom_nv4x_adl/v1.6.0/novacustom_nv4x_adl_ec_v1.6.0.rom
flashrom -p internal --ifd -i bios -w main.rom && system76_ectool flash ec.rom
```

**Comment by macpijan** [comment]
Hi @geraldhobbes and thanks for the report. While we do not have this particular dock to test, we will be looking into some USB-C this week. We will let you know here if we have any update in this area.

**Comment by macpijan** [comment]
@geraldhobbes Station resetting sounds like the laptops wants to pull more power than the station can provide.

Can you please provide also details on which power supply do you have? Is it 90W, 130W, or another one? Do you have ability to try with another one?

**Comment by mkopec** [comment]
Yes, it sounds like overcurrent protection engaging in the dock. The laptop is not applying the correct power limits. It's possible that we need to fix the charger current logic to account for different power supplies. I can reproduce with Lenovo ThinkPad 65W and with a Baseus 65W chargers.

**Comment by geraldhobbes** [comment]
> Can you please provide also details on which power supply do you have? Is it 90W, 130W, or another one? Do you have ability to try with another one (more power)?

Dell's [user guide](https://dl.dell.com/topicspdf/dell-wd19-130w-dock_users-guide_en-us.pdf) says about power delivery: “Non-Dell systems: 90 W support for USB Power Delivery 3.0”. The label of the Dell dock reads 19.5V 6.7A/ 9.23A. The label of the AC adapter plugged into the dock reads 180W 19.5V 9.23A.

**Comment by wessel-novacustom** [comment]
@mkopec was able to reproduce the problem under Ubuntu with a Lenovo 65W USB-C charger, while stress testing the CPU and toggling the fan speed to 100%. We have a test EC binary with a potential fix and I will try to flash that version. I will let you know the result as quickly as possible.

**Comment by wessel-novacustom** [comment]
@geraldhobbes I can confirm that, with the new EC firmware binary, the problem has been solved.

Would you like to help us and others by testing it in your use case scenario with your specific dock?

If so, please update to Dasharo v1.7.1 according to the [firmware update documentation page](https://docs.dasharo.com/unified/novacustom/firmware-update/) and let me know, I will then send you the instructions how to apply the [test EC binary](https://github.com/Dasharo/ec/actions/runs/6862979426?pr=53).

**Comment by geraldhobbes** [comment]
Tomorrow, I might have time to try flashing that firmware

**Comment by wessel-novacustom** [comment]
@geraldhobbes That would be very helpful. Do you need flashing instructions?

**Comment by geraldhobbes** [comment]
So, I let DTS update to the latest version, v1.7.1. Then, I opened the DTS shell and ran `system76_ectool flash novacustom_nv41x_adl_ec.rom`. After rebooting, `system76_ectool info` showed `version: 2023-11-14_9a7058c`. Unfortunately, the Dell dock doesn't charge the laptop and it malfunctions in the same way. Did I flash it correctly?

**Comment by wessel-novacustom** [comment]
@geraldhobbes Yes, you flashed it correctly.

The fix was particularly created for PowerDelivery input only, where even lower power inputs over PowerDelivery should work better now.

Maybe the laptop is confused as both PowerDelivery is active and the regular AC adaptor was attached already.

So what happens if you:

1. Shut down the laptop.
2. Remove the AC adaptor and docking station.
3. Plug in the docking station.
4. Try again to reproduce the problem.

- Does the laptop charge correctly without interruptions (already; without the regular AC adaptor, but only PowerDelivery being active)?
- Does the bug happen in this case?

@mkopec Do you suggest to use this binary (`version: 2023-11-14_9a7058c`) or rather another EC binary? Also, the problem must be purely EC related, right?

**Comment by geraldhobbes** [comment]
If I boot with only the docking station plugged in, it also malfunctions.

**Comment by GoodGuyPat** [comment]
I can replicate this issue on a Novacustom NS50_MU, Dasharo v1.5, on archlinux 6.6.1-arch1-1, docking station Dell WD15. I normally use PD exclusively (without the AC adaptor plugged in), and that worked stably with Dasharo v1.4.

I have now plugged the AC adaptor and the issue is no longer present.

**Comment by wessel-novacustom** [comment]
@mkopec There still seems to be a problem with the Embedded Controller (EC) firmware. I could reproduce the same problem with the WAVLINK WL-UG69PD2 Pro 100W PD DisplayLink docking station. That dock shows the problems way more often  than the DisplayAltMode docks you are currently using. I have therefore sent you this docking station, it should arrive on Thursday.

**Comment by wessel-novacustom** [comment]
The problem seems to be the best reproducible with a low battery charge (< 50%) and with a docking station that has Power Delivery of up to 65W, while the energy consumption of the laptop and the CPU load is high.

**Comment by mkopec** [comment]
Some chargers have more aggressive overcurrent protection than others. We'll try tuning the throttling parameters to ensure that OCP is never tripped unnecesarily.

**Comment by wessel-novacustom** [comment]
@mkopec That would be great, please let me know once you have a test binary.

**Comment by mkopec** [comment]
@wessel-novacustom Updated binaries for testing should be generated by CI in PR: https://github.com/Dasharo/ec/pull/55

Tested on NV4x ADL with a Baseus 65W charger, now I can't trigger OCP anymore in 65W and 45W modes. Need to also test other models without HPB to verify that the power limits work there as well.

**Comment by wessel-novacustom** [comment]
@mkopec Thank you. I flashed the provided test EC binary externally and indeed, this particular problem does no longer happen, even not for our DisplayLink docking station that only supplies up to 65W (PD).

That being said, I found three new problems for this EC version:

1. When the laptop (NV40MZ-i5) was fully charged and attached to the Wavlink Rev. e dock, I tried to suspend with S0ix set in Windows, the laptop showed some weird behaviour. The screen went off, the keyboard illumination went off, then after a few seconds, the charging LED indicator changed from green to off, to orange and back to green again. This also caused the screen display to go on again as well as the keyboard illumination. After a few seconds, the screen went off again and after a few more seconds (or maybe even a minute), the keyboard illumination shut off and the power LED was blinking green, while the battery LED indicator was green as well. Then when attempting to wake up from suspend mode, the laptop refused: a keypress didn't bring the laptop back to life, nor did a power button press. I hard to hold the power button to do a hard reset to make the laptop working again.
2. In Windows on the same device, when removing the dock and then immediately try to shut down the laptop, the laptop hangs at the shutdown process when the text _Shutting down_ is showed.
3. The [video initialisation](https://github.com/Dasharo/dasharo-issues/issues/560) problems are back, possibly because the EC is not re-initialising the battery charging state after shutting down.
4. When hotplugging the Wavlink DisplayLink (65W) dock, the battery icon turned orange indicating that it is charging, however, Ubuntu was indicating that it was still running on battery mode. Only after 1 - 2 minutes, the battery icon in Ubuntu was telling me that the battery was charging. This seems to be a Ubuntu bug however, rather than a firmware bug.
